utils/resolve_weather_station_id.py

```diff
@@ -0,0 +1,759 @@
+    if isinstance(weather_station, dict) and "Name" in weather_station:
+        return weather_station["Name"]
+    elif isinstance(weather_station, list):
+        for item in weather_station:
+            if isinstance(item, dict) and "Name" in item:
+                return item["Name"]
+    return None
+
+
+def _search_dict_for_weather_station(data, path=""):
+    """Search dictionary for weather station."""
+    if "WeatherStation" in data:
+        weather_station = data["WeatherStation"]
+        name = _extract_name_from_weather_station(weather_station)
+        if name:
+            return name
+
+    # Recursively search all values in the dict
+    for key, value in data.items():
+        result = find_weather_station_name(value, f"{path}.{key}" if path else key)
+        if result:
+            return result
+    return None
+
+
+def _search_list_for_weather_station(data, path=""):
+    """Search list for weather station."""
+    for i, item in enumerate(data):
+        result = find_weather_station_name(item, f"{path}[{i}]" if path else f"[{i}]")
+        if result:
+            return result
+    return None
+
+
+def find_weather_station_name(data, path=""):
+    """
+    Recursively search for WeatherStation tag and extract the Name value.
+
+    Args:
+        data: The data structure to search (dict, list, or primitive)
+        path: Current path for debugging (optional)
+
+    Returns:
+        str: Weather station name if found, None otherwise
+    """
+    if isinstance(data, dict):
+        return _search_dict_for_weather_station(data, path)
+    elif isinstance(data, list):
+        return _search_list_for_weather_station(data, path)
+
+    return None
+
+
+if __name__ == "__main__":
+    start_time = time.time()
+    product = "resstock"
+    release_year = "2022"
+    weather_file = "amy2012"
+    release_version = "1"
+    state = "NY"
+    upgrade_id = "0"
+
+    # You can customize the update intervals and parallel workers:
+    # status_update_interval_seconds: How often to print status updates (in seconds)
+    # status_update_interval_bldgs: Force status update every N buildings
+    # max_workers: Number of parallel threads for downloading
+    bldg_ids = fetch_bldg_ids(product, release_year, weather_file, release_version, state, upgrade_id)
+    df = resolve_weather_station_id(
+        bldg_ids,
+        product,
+        release_year,
+        weather_file,
+        release_version,
+        state,
+        upgrade_id,
+        status_update_interval_seconds=15,  # Update every 15 seconds
+        status_update_interval_bldgs=25,  # Also update every 25 buildings
+        max_workers=20,  # Use 20 parallel threads for downloading
+    )
+
+    # Create output directory if it doesn't exist
+    output_dir = "buildstock_fetch/data/weather_station_map"
+    os.makedirs(output_dir, exist_ok=True)
+
+    # Remove duplicates and sort by building ID for better organization
+    if "bldg_id" in df.columns:
```

Comment:
Is `bldg_id` always supposed to be here? If so, use an `assert`, not a conditional.

Reply:
Also, this whole section should be a function.